### DISCUSSION TITLE
transfer file from nautilus & nemo & thunar right-click menu?

### DISCUSSION THREAD

**Discussion by Aleksanaa** [andyholmes/valent#355] [Help] [answered]
Any command to specify the file to transfer?

**Comment by andyholmes** [comment, marked as answer]
I don't have any plans at this time to write a Nautilus/Nemo extension, but it would fairly simple to adapt the one from [GSConnect](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/nautilus-extension/nautilus-gsconnect.py). Assuming Thunar plugins work in the same way, that wouldn't be too hard to port to.

Personally, I just use the *Open With...* dialog, which pops open a device selection window:

![image](https://user-images.githubusercontent.com/1265208/209410310-d62ddc9a-5089-4348-8f5f-0f9afc5152aa.png)

**Reply by andyholmes** [reply]
By the way, if you want any help with porting that Nautilus/Nemo extension let me know and I'd be happy to.

Writing a simple CLI to do this would work basically the same way, so I can help with that too if you like.

I'll write a full CLI at some point, I just need to figure out how to allow for plugins to work with that.

**Reply by Aleksanaa** [reply]
> I don't have any plans at this time to write a Nautilus/Nemo extension, but it would fairly simple to adapt the one from [GSConnect](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/nautilus-extension/nautilus-gsconnect.py). Assuming Thunar plugins work in the same way, that wouldn't be too hard to port to.
> 
> Personally, I just use the _Open With..._ dialog, which pops open a device selection window:
> 
> ![image](https://user-images.githubusercontent.com/1265208/209410310-d62ddc9a-5089-4348-8f5f-0f9afc5152aa.png)

Ha I didn't notice 'open with', but I think this can be the solution if no other command line arguments implemented.

Then I can use this:
https://unix.stackexchange.com/questions/124811/is-there-an-open-with-command-for-the-command-line

**Reply by andyholmes** [reply]
FWIW, I think you can also use `gapplication launch ca.andyholmes.Valent filename.ext`. I think it should be able to handle multiple files too, so you might be able to make a script like:

```sh
#!/bin/sh -e

gapplication launch ca.andyholmes.Valent "${@}";
```

And call it like `./share.sh file1.txt file2.txt`.

**Comment by DodoLeDev** [comment]
### Hi everyone!
**I come to you with excellent news:** I've finalized a first version of a Nautilus extension, which enables you to send files to your devices from a simple right-click entry!

![image](https://github.com/andyholmes/valent/assets/59477313/40e79ae1-4581-4d26-909c-62cba5ba440d)

The extension is written in Python thanks to the `natilus-python` API. So you'll need to install it for the extension to work.

I based my work on [another Nautilus extension](https://github.com/GeopJr/Collision/blob/main/nautilus-extension/collision-extension.py) I developed for another GTK4 project: [Collision](https://github.com/GeopJr/Collision) (feel free to discover and/or contribute to this other Linux application)

_Tell me if it works for you, and have fun with it!_

## Download the extension here:
<a href="https://github.com/andyholmes/valent/files/12009346/valent-share.txt"><img width=130px src="https://img.shields.io/badge/valent--share.py-167460?style=flat-square&logo=python&logoColor=white"></a>

1. Rename the file from .txt to .py
2. Move it to `~/.local/share/nautilus-python/extensions/` (create directories if they don't exist yet)
3. Force quit nautilus with `nautilus -q`
4. Restart Nautilus
5. Enjoy

**Reply by andyholmes** [reply]
Very cool, thanks for sharing!

**Reply by ArnavGhosh2327** [reply]
Sorry, but it does not work by default. It is doing exactly the same thing when I invoke “valent filename” from terminal, where it won't actually send the files. Valent pretty sure does not have permission to access those files, which is making it hard to share files, so maybe the Flatpak users might need to actually Change the Permissions manually using Flatpak Permission manager.
![2025-06-05T14:52:12,700815946+05:30](https://github.com/user-attachments/assets/af055f02-5c75-41cc-ba88-6b49d60819a8)

**Reply by ArnavGhosh2327** [reply]
UPDATE: It works after changing the Permissions

**Reply by andyholmes** [reply]
> Valent pretty sure does not have permission to access those files, which is making it hard to share files, so maybe the Flatpak users might need to actually Change the Permissions manually using Flatpak Permission manager.

Ah, thanks for checking back about this. The file passing (e.g. `valent %U`) uses the `org.freedesktop.Application.Open()` method, so I assumed the desktop would automatically handle access to files. I'll try and have a look and see what can be done about that.
